### ISSUE TITLE
Type of destructured arguments are unknown with `chain.from_iterable()`

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#6455]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from itertools import chain

times = [
    f'{12 if hour == 0 else hour!s}:{minute!s:0>2} {meridian}'
    for hour, minute, meridian in chain.from_iterable(
        chain.from_iterable(
            [(hour, minute, meridian) for minute in range(0, 60, 15)]
            for hour in range(12)
        )
        for meridian in ('am', 'pm')
    )
]
```

The following errors are generated in strict mode:

```
blah.py
  blah.py:7:9 - error: Type of "hour" is unknown (reportUnknownVariableType)
  blah.py:7:15 - error: Type of "minute" is unknown (reportUnknownVariableType)
  blah.py:7:23 - error: Type of "meridian" is unknown (reportUnknownVariableType)
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool?
Command line

Which version?
1.1.336

**Comment by bryanforbes** [comment]
@juk-r thank you for the shorter version of what I was doing. However the original code I shared here should still type check correctly.

**Comment by erictraut** [comment]
Thanks for the bug report. This was due to a long-standing bug that was exposed by a recent (seemingly unrelated) change that I made. It will be addressed in the next release. I've added a variant of your code snippet to the pyright unit tests to help prevent future regressions.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
